File: src/cpp/ee/ads/internal/DefaultFullScreenAd.cpp
//
//  DefaultFullScreenAd.cpp
//  Pods
//
//  Created by eps on 1/26/21.
//

#include "ee/ads/internal/DefaultFullScreenAd.hpp"

#include <cassert>

#include <ee/core/Delay.hpp>
#include <ee/core/ILogger.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Utils.hpp>
#include <ee/nlohmann/json.hpp>

#include "ee/ads/internal/AsyncHelper.hpp"
#include "ee/ads/internal/Capper.hpp"

namespace ee {
namespace ads {
using Self = DefaultFullScreenAd;

Self::DefaultFullScreenAd(
    const std::string& prefix, IMessageBridge& bridge, ILogger& logger,
    const std::shared_ptr<IAsyncHelper<AdResult>>& displayer,
    const Destroyer& destroyer, const ResultParser& resultParser,
    const std::string& network, const std::string& adId)
    : prefix_(prefix)
    , bridge_(bridge)
    , logger_(logger)
    , displayer_(displayer)
    , destroyer_(destroyer)
    , resultParser_(resultParser)
    , network_(network)
    , adId_(adId)
    , messageHelper_(prefix, adId) {
    logger_.debug("%s: prefix = %s id = %s", __PRETTY_FUNCTION__,
                  prefix_.c_str(), adId_.c_str());
    loadCapper_ = std::make_shared<Capper>(30);
    loader_ = std::make_unique<AsyncHelper<bool>>();

    bridge_.registerHandler(
        [this](const std::string& message) { //
            onLoaded();
        },
        messageHelper_.onLoaded());
    bridge_.registerHandler(
        [this](const std::string& message) { //
            auto json = nlohmann::json::parse(message);
            onFailedToLoad(json["code"], json["message"]);
        },
        messageHelper_.onFailedToLoad());
    bridge_.registerHandler(
        [this](const std::string& message) {
            auto json = nlohmann::json::parse(message);
            onFailedToShow(json["code"], json["message"]);
        },
        messageHelper_.onFailedToShow());
    bridge_.registerHandler(
        [this](const std::string& message) { //
            onClicked();
        },
        messageHelper_.onClicked());
    bridge_.registerHandler(
        [this](const std::string& message) { //
            auto result = resultParser_(message);
            onClosed(result);
        },
        messageHelper_.onClosed());
}

Self::~DefaultFullScreenAd() = default;

void Self::destroy() {
    logger_.debug("%s: prefix = %s id = %s", __PRETTY_FUNCTION__,
                  prefix_.c_str(), adId_.c_str());

    bridge_.deregisterHandler(messageHelper_.onLoaded());
    bridge_.deregisterHandler(messageHelper_.onFailedToLoad());
    bridge_.deregisterHandler(messageHelper_.onFailedToShow());
    bridge_.deregisterHandler(messageHelper_.onClicked());
    bridge_.deregisterHandler(messageHelper_.onClosed());

    destroyer_();
}

bool Self::isLoaded() const {
    auto response = bridge_.call(messageHelper_.isLoaded());
    return core::toBool(response);
}

Task<bool> Self::load() {
    logger_.debug("%s: prefix = %s id = %s loading = %s", __PRETTY_FUNCTION__,
                  prefix_.c_str(), adId_.c_str(),
                  core::toString(loader_->isProcessing()).c_str());
    if (loadCapper_->isCapped()) {
        co_return false;
    }
    loadCapper_->cap();
    auto result = co_await loader_->process(
        [this] { //
            bridge_.call(messageHelper_.load());
        },
        [](bool result) {
            // OK.
        });
    co_return result;
}

Task<AdResult> Self::show() {
    logger_.debug("%s: prefix = %s id = %s displaying = %s",
                  __PRETTY_FUNCTION__, prefix_.c_str(), adId_.c_str(),
                  core::toString(displayer_->isProcessing()).c_str());
    auto result = co_await displayer_->process(
        [this] { //
            bridge_.call(messageHelper_.show());
        },
        [](AdResult result) {
            // OK.
        });
    co_return result;
}

void Self::onLoaded() {
    logger_.debug("%s: prefix = %s id = %s loading = %s", __PRETTY_FUNCTION__,
                  prefix_.c_str(), adId_.c_str(),
                  core::toString(loader_->isProcessing()).c_str());
    if (loader_->isProcessing()) {
        loader_->resolve(true);
        dispatchEvent([this](auto&& observer) {
            if (observer.onLoadResult) {
                observer.onLoadResult({
                    .network = network_,
                    .result = true,
                });
            }
        });
    } else {
        logger_.error("%s: this ad is expected to be loading",
                      __PRETTY_FUNCTION__);
        assert(false);
    }
    dispatchEvent([](auto&& observer) {
        if (observer.onLoaded) {
            observer.onLoaded();
        }
    });
}

void Self::onFailedToLoad(int code, const std::string& message) {
    logger_.debug("%s: prefix = %s id = %s loading = %s message = %s",
                  __PRETTY_FUNCTION__, prefix_.c_str(), adId_.c_str(),
                  core::toString(loader_->isProcessing()).c_str(),
                  message.c_str());
    if (loader_->isProcessing()) {
        loader_->resolve(false);
        dispatchEvent([this, code, message](auto&& observer) {
            if (observer.onLoadResult) {
                observer.onLoadResult({
                    .network = network_,
                    .result = false,
                    .errorCode = code,
                    .errorMessage = message,
                });
            }
        });
    } else {
        logger_.error("%s: this ad is expected to be loading",
                      __PRETTY_FUNCTION__);
        assert(false);
    }
}

void Self::onFailedToShow(int code, const std::string& message) {
    logger_.debug("%s: prefix = %s id = %s displaying = %s message = %s",
                  __PRETTY_FUNCTION__, prefix_.c_str(), adId_.c_str(),
                  core::toString(displayer_->isProcessing()).c_str(),
                  message.c_str());
    if (displayer_->isProcessing()) {
        displayer_->resolve(AdResult::Failed);
    } else {
        logger_.error("%s: this ad is expected to be displaying",
                      __PRETTY_FUNCTION__);
        assert(false);
    }
}

void Self::onClicked() {
    logger_.debug("%s: prefix = %s id = %s", __PRETTY_FUNCTION__,
                  prefix_.c_str(), adId_.c_str());
    dispatchEvent([](auto&& observer) {
        if (observer.onClicked) {
            observer.onClicked();
        }
    });
}

void Self::onClosed(AdResult result) {
    logger_.debug("%s: prefix = %s id = %s displaying = %s",
                  __PRETTY_FUNCTION__, prefix_.c_str(), adId_.c_str(),
                  core::toString(displayer_->isProcessing()).c_str());
    if (displayer_->isProcessing()) {
        displayer_->resolve(result);
    } else {
        logger_.error("%s: this ad is expected to be displaying",
                      __PRETTY_FUNCTION__);
        assert(false);
    }
}
} // namespace ads
} // namespace ee

File: src/ts/src/ads/internal/DefaultFullScreenAd.ts
import {
    ILogger,
    IMessageBridge,
    ObserverManager,
    Utils,
} from "../../core";
import { AdObserver } from "../IAd";
import {
    AdResult,
    IFullScreenAd,
} from "../IFullScreenAd";
import { AsyncHelper } from "./AsyncHelper";
import { Capper } from "./Capper";
import { IAsyncHelper } from "./IAsyncHelper";
import { ICapper } from "./ICapper";
import { MessageHelper } from "./MessageHelper";

type Destroyer = () => void;
type ResultParser = (message: string) => AdResult;

export class DefaultFullScreenAd extends ObserverManager<AdObserver> implements IFullScreenAd {
    private readonly kTag = `DefaultFullScreenAd`;
    private readonly _prefix: string;
    private readonly _bridge: IMessageBridge;
    private readonly _logger: ILogger;
    private readonly _displayer: IAsyncHelper<AdResult>;
    private readonly _destroyer: Destroyer;
    private readonly _resultParser: ResultParser;
    private readonly _network: string;
    private readonly _adId: string;
    private readonly _messageHelper: MessageHelper;
    private readonly _loadCapper: ICapper;
    private readonly _loader: IAsyncHelper<boolean>;

    public constructor(
        prefix: string,
        bridge: IMessageBridge,
        logger: ILogger,
        displayer: IAsyncHelper<AdResult>,
        destroyer: Destroyer,
        resultParser: ResultParser,
        network: string,
        adId: string) {
        super();
        this._prefix = prefix;
        this._bridge = bridge;
        this._logger = logger;
        this._displayer = displayer;
        this._destroyer = destroyer;
        this._resultParser = resultParser;
        this._network = network;
        this._adId = adId;
        this._messageHelper = new MessageHelper(prefix, adId);
        this._loadCapper = new Capper(30);
        this._loader = new AsyncHelper<boolean>();

        this._logger.debug(`${this.kTag}: constructor: prefix = ${this._prefix} id = ${this._adId}`);
        this._bridge.registerHandler(_ => this.onLoaded(), this._messageHelper.onLoaded);
        this._bridge.registerHandler(message => {
            const json: {
                code: number,
                message: string,
            } = JSON.parse(message);
            this.onFailedToLoad(json.code, json.message);
        }, this._messageHelper.onFailedToLoad);
        this._bridge.registerHandler(message => {
            const json: {
                code: number,
                message: string,
            } = JSON.parse(message);
            this.onFailedToShow(json.code, json.message);
        }, this._messageHelper.onFailedToShow);
        this._bridge.registerHandler(_ => this.onClicked(), this._messageHelper.onClicked);
        this._bridge.registerHandler(message => {
            const result = this._resultParser(message);
            this.onClosed(result);
        }, this._messageHelper.onClosed);
    }

    public destroy(): void {
        this._logger.debug(`${this.kTag}: destroy: prefix = ${this._prefix} id = ${this._adId}`);
        this._bridge.deregisterHandler(this._messageHelper.onLoaded);
        this._bridge.deregisterHandler(this._messageHelper.onFailedToLoad);
        this._bridge.deregisterHandler(this._messageHelper.onFailedToShow);
        this._bridge.deregisterHandler(this._messageHelper.onClicked);
        this._bridge.deregisterHandler(this._messageHelper.onClosed);
        this._destroyer();
    }

    public get isLoaded(): boolean {
        const response = this._bridge.call(this._messageHelper.isLoaded);
        return Utils.toBool(response);
    }

    public async load(): Promise<boolean> {
        this._logger.debug(`${this.kTag}: load: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing}`);
        if (this._loadCapper.isCapped) {
            return false;
        }
        this._loadCapper.cap();
        return await this._loader.process(
            () => this._bridge.call(this._messageHelper.load),
            result => {
                // OK.
            });
    }

    public show(): Promise<AdResult> {
        this._logger.debug(`${this.kTag}: show: prefix = ${this._prefix} id = ${this._adId} displaying = ${this._displayer.isProcessing}`);
        return this._displayer.process(
            () => this._bridge.call(this._messageHelper.show),
            result => {
                // OK.
            });
    }

    private onLoaded(): void {
        this._logger.debug(`${this.kTag}: onLoaded: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing}`);
        if (this._loader.isProcessing) {
            this._loader.resolve(true);
            this.dispatchEvent(observer => observer.onLoadResult && observer.onLoadResult({
                network: this._network,
                result: true,
                errorCode: 0,
                errorMessage: ``,
            }));
        } else {
            // Assert.
        }
        this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
    }

    private onFailedToLoad(code: number, message: string): void {
        this._logger.debug(`${this.kTag}: onFailedToLoad: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing} message = ${message}`);
        if (this._loader.isProcessing) {
            this._loader.resolve(false);
            this.dispatchEvent(observer => observer.onLoadResult && observer.onLoadResult({
                network: this._network,
                result: true,
                errorCode: code,
                errorMessage: message,
            }));
        } else {
            // Assert.
        }
    }

    private onFailedToShow(code: number, message: string): void {
        this._logger.debug(`${this.kTag}: onFailedToLoad: prefix = ${this._prefix} id = ${this._adId} displaying = ${this._displayer.isProcessing} message = ${message}`);
        if (this._displayer.isProcessing) {
            this._displayer.resolve(AdResult.Failed);
        } else {
            // Assert.
        }
    }

    private onClicked(): void {
        this._logger.debug(`${this.kTag}: onClicked: prefix = ${this._prefix} id = ${this._adId}`);
        this.dispatchEvent(observer => observer.onClicked && observer.onClicked());
    }

    private onClosed(result: AdResult): void {
        this._logger.debug(`${this.kTag}: onClosed: prefix = ${this._prefix} id = ${this._adId} displaying = ${this._displayer.isProcessing}`);
        if (this._displayer.isProcessing) {
            this._displayer.resolve(result);
        } else {
            // Assert..
        }
    }
}